src/ThemeProvider.tsx

```diff
@@ -83,8 +84,13 @@ export const ThemeProvider: React.FC<React.PropsWithChildren<ThemeProviderProps>
         // if the resolved color mode passed on from the server is not the resolved color mode on client, change it!
         if (resolvedColorModePassthrough.current !== resolvedColorModeOnClient) {
           window.setTimeout(() => {
-            // override colorMode to whatever is resolved on the client to get a re-render
-            setColorMode(resolvedColorModeOnClient)
+            // use ReactDOM.flushSync to prevent automatic batching of state updates since React 18
+            // ref: https://github.com/reactwg/react-18/discussions/21
+            ReactDOM.flushSync(() => {
```

Comment:
Just checking, would this lead to an error with React 17?

Reply:
https://github.com/facebook/react/pull/10225
https://github.com/facebook/react/issues/11527#issuecomment-360199710

It seems that `ReactDOM.flushSync(batch)` is already introduced before React 17.